Release action

poshul opened the pull request:
## Description
This unfortunately needs to have "release" in the branch name to work. 

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

poshul committed:
Start adding the Release action Jenkins stuff to GHA

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

Regex bs

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

Add breadcrumbs to end of file for planned work

jpfeuffer commented:
Have you considered the dis-/advantages of having the release job separately from the packaging?

I think both would work.

Do you want to trigger on release tag or manual trigger only and the job tags the repo?

poshul commented:
I've thought about having separate jobs. It did work from the Jenkins side.
My thoughts are that actually doing a release still requires manual
interaction, (sending out an email, making sure that cython built etc.) and
so the manual trigger route is the way to go, and if we are triggering it
manually anyway, just having it as a single action means that we don't have
to go and fetch SHA's from the archives, or double check that the artifacts
are still from HEAD on the release branch. I'm open to discussion on it
though.


On Fri, Nov 17, 2023 at 7:57 PM Julianus Pfeuffer ***@***.***>
wrote:

> Have you considered the dis-/advantages of having the release job
> separately from the packaging?
>
> I think both would work.
>
> Do you want to trigger on release tag or manual trigger only and the job
> tags the repo?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMS/OpenMS/pull/7210#issuecomment-1816934120>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABD3C4LS7NISMFE7ZLYNZTYE6XTBAVCNFSM6AAAAAA7P3FNQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMJWHEZTIMJSGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jpfeuffer commented:
Sounds reasonable. What do you mean with "Cython built"?
Would be cool to just check the status of the last actions with GH cli. I'm pretty sure it is possible.

jpfeuffer commented:
Another question: do you have a plan for integrating the macOS Silicon builds?

Options:
1. cross-compile without tests on GHA
2. self-hosted runner on GHA
3. stay with current setup
4. Jenkins with Jenkinsfile

I think 2 would be worthwhile, even when switching once Silicon builds become available for open source (probably mid 2024?). If that is out of scope I would probably stay with the current setup and try to integrate the artifacts from Jenkins as smoothly as possible.

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul commented:
> Sounds reasonable. What do you mean with "Cython built"? Would be cool to just check the status of the last actions with GH cli. I'm pretty sure it is possible.

Don't we still build Cython through: https://abibuilder.cs.uni-tuebingen.de/jenkins/job/openms/job/ntly/job/conda/?

jpfeuffer commented:
Ah you mean conda?
Multiple options:
- Query jenkins for success via its API: https://abibuilder.cs.uni-tuebingen.de/jenkins/job/openms/job/ntly/job/conda/lastBuild/api (for example [xml-output](https://abibuilder.cs.uni-tuebingen.de/jenkins/job/openms/job/ntly/job/conda/lastBuild/api/xml); you need cookies/be logged in or get an API key for programmatic access).
- Move to conda-forge or talk to bioconda people about auto-update of the recipe based on GitHub-Releases. Anything that cannot be automated, like changes in build.sh of the recipe must be made with a PR to bioconda(-forge) after release (potentially automated with GH CLI). Actually, thinking about it, maybe we should move the build.sh(s) for conda into the OpenMS source repository and make conda run this file instead; then it might be updated with the release automatically. Never thought about it (I might be wrong).
- Move the conda job to GH Actions. It's all docker-based, should be rather easy (except for the workarounds needed for the year long bugs). Runtime might be an issue (probably not). Could be alleviated by:
  - reducing the python versions
  - splitting pyopenms from openms (should be working in theory already)
  - using our docker builder as a selfhosted runner on GHA

jpfeuffer commented:
The update of the bioconda(-forge) recipe (=item 2) is actually kind of unrelated to our internal "nightly" conda builds but must be considered.

poshul committed:
Added logic to update SHAs

poshul committed:
Further work on release

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
comment out inprogress code

poshul committed:
regex

poshul committed:
more regex

poshul committed:
fix path for changelog

poshul committed:
fix name of changelog artifact

poshul committed:
comment out windows and mac builds to speed debugging

poshul committed:
fix changelog name in releasetext

poshul committed:
re-enable slow OS'

poshul committed:
fix path

poshul committed:
handle rrsync competition for lock.

poshul committed:
fix other path

poshul committed:
bash indent bs

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -456,6 +507,53 @@ jobs:
+            ${{ github.workspace }}/*.exe
+            ${{ github.workspace }}/*.deb
+            ${{ github.workspace }}/*.pkg
+            # TODO add src
```
In theory a src gz is always available at the release page, separately from the artifacts but we are currently adding a modfied src.gz with docs, with checked out contrib (and Thirdparty?) and without .git index.

poshul replied:
Okay I'm fine with leaving it out of here then.

jpfeuffer replied:
I didn't directly mean that 😅
Bioconda might use that artifact but maybe there are better ways and this artifact is indeed not needed.

(We probably should use the autogenerated artifact anyway when switching to Autoupdates to conda based on GitHub releases)

poshul replied:
Ah, okay. I can look into that after I get the current batch of changes to work. I have to say, I hadn't realized just how much patience it takes to get all the CI to work. Props to you for all of your work on it over the years.

poshul committed:
fix release syntax, fix bash stuff.

poshul committed:
create a tag first, because the release action doesn't do so.

poshul committed:
replace release action

poshul committed:
try with a static tag

poshul committed:
fix regex for uploadpath, mark release as Draft

poshul committed:
fixed locking problem using changes to authorized_keys on archive host. Revert attempts to do so in here

poshul committed:
fix regex for upload path in KNIME addons

poshul committed:
comma delimit release paths.

poshul committed:
fix (hopefully) getting the RUN_NAME for release.

poshul committed:
fix envs in create_SHA

poshul committed:
update links to latest, merge-to-develop

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -547,14 +552,23 @@ jobs:
-          tag: 2.9.1-test #${{ steps.extract_branch.outputs.RUN_NAME }}
-          draft: true
+          tag: ${{ env.RUN_NAME }}
+          draft: true #FIXME revert
```
Draft might actually be fine. Could be a good way to double-check. But on the other hand, if something is wrong, the damage was done already, so it probably does not matter.

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -889,17 +910,34 @@ jobs:
+        with:
+          script: |
+            const { repo, owner } = context.repo;
+            const result = await github.rest.pulls.create({
```
Hmm I dont think you need a PR. release branches should ALWAYS be ff-only mergeable into main/master. If not, you should not even start the release ;) Maybe check this in the very beginning and maybe also tag the merged commit instead of the head commit of the release branch (although it might be that the head commit carries over its tags when you merge it ff-only into master/main).
What would be nice, is to have a PR for master or release to be merged back into develop, though!

poshul committed:
fix latest linking

poshul committed:
make tags using gh instead of javascript

poshul committed:
fix bash bs

poshul committed:
fix supernumerary "

poshul committed:
get rid of unnecessary ref

poshul committed:
explicitly specify API token

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
fix	o

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -488,10 +600,16 @@ jobs:
           # Upload documentation FIXME
           rsync --progress -avz -e "ssh -i ~/.ssh/private.key -o StrictHostKeyChecking=no" $GITHUB_WORKSPACE/docs/ "$USER@$HOST:/Documentation/${folder}"
+
+          if [[ 'true' == ${{ inputs.mark_as_latest }} ]]; then
```
Can we do an initial check if a release is really justified:
do_release && branch=~ release(0-9) && BASH_REMATCH matches OpenMS version in git

Then fail early? Otherwise this will all execute and you will link anything to latest

jpfeuffer replied:
MAybe you can add it to the changelog extraction step and move it up a little. Something without too much effort.

poshul replied:
https://github.com/OpenMS/OpenMS/pull/7210/commits/04ce8680bbdbdbe48740509a1fa75f0ba89aa064 does what I understand you to be asking, can you double check that this is what you meant?

jpfeuffer replied:
yep, very cool!

poshul committed:
fix typo notes for monday

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
move changelog creation up and add checks for nonsensical inputs. Fail if trying to mark_latest on a version that already exists

poshul committed:
Merge branch 'release/2.9.1-action' of ssh://github.com/OpenMS/OpenMS into release/2.9.1-action

poshul committed:
fix spacing

poshul committed:
fix ref to removed tool

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -131,24 +143,71 @@ jobs:
+          echo $END
+          awk "NR > $START && NR < $END" OpenMS/CHANGELOG > ${{ github.workspace }}/changelog.txt
+
+          # check if we want to be marked as latest, and if we actually should.
```
Cool. Really thorough checks!

poshul committed:
correct how we get the shas

poshul committed:
fixed (hopefully) tag syntax

poshul committed:
Fix syntax for gh

poshul committed:
Update openms_ci_matrix_full.yml

explicitly add write permissions to github.token

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul commented:
@jpfeuffer I'm having huge issues getting the tags created. It works fine on my PC using a personal access token, but I get permission errors using the GITHUB_TOKEN, even when I've explicitly given it permissions to write. Do you have any insight as to the permission limits on that token, and whether I just need to bite the bullet and create an explicit token for this workflow?

jpfeuffer commented:
The GitHub_Token is scoped to the triggering repo. You need to use an app token e.g. from our OpenMSGitHubApp

jpfeuffer commented:
```
# get token
      - uses: actions/create-github-app-token@v1
        id: app-token
        with:
          app-id: ${{ vars.OPENMS_GITHUB_APP_ID }}
          private-key: ${{ secrets.OPENMS_GITHUB_APP_PRIVATE_KEY }}
          owner: ${{ github.repository_owner }}

# example use
      - uses: peter-evans/create-or-update-comment@v3
        with:
          token: ${{ steps.app-token.outputs.token }}
          issue-number: ${{ github.event.issue.number }}
          body: "Hello, World!"

```

poshul committed:
Update openms_ci_matrix_full.yml

Add explicit token

poshul committed:
Merge branch 'develop' into release/2.9.1-action

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -675,3 +816,109 @@ jobs:
+        name: create tags for other repos
+        shell: bash
+        env: 
+          GH_TOKEN: ${{ github.event.issue.number }}
```
I guess this needs to be changed to use the new token.

jpfeuffer replied:
Why issue number, though?
Shouldn't it be ${{ steps.app-token.outputs.token }}

poshul committed:
Update openms_ci_matrix_full.yml

Fix bash issue when running as a push

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

refactor where we check inputs to avoid bash issues. Fix assignment instead of equality test in do_package. Export version number from changelog step

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

why the heck is it workflow_run when the value is workflow_dispatch?

poshul committed:
Update openms_ci_matrix_full.yml

timosachsenberg commented:
any update here?

timosachsenberg commented:
/review

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
Update openms_ci_matrix_full.yml

Fix duplicate ID from merge

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

GHA parser utterly fails to determine where if finds a syntax error. Fixed syntax error

poshul committed:
fix conditionals for mark_as_latest to only run if we are a workflow_dispatch

poshul committed:
Break out linking as latest

bash chokes on variables that may or may not exist (like inputs.mark_as_latest) refactor the behavior of mark_as_latest to avoid this.

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

fix two more sha issues

poshul committed:
Update openms_ci_matrix_full.yml

trying to not specify python3 version

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

force GHA to not update preinstalled packages in brew

poshul commented:
Tagging now works, aside from the two GenericWorkflowNodes repos. @jpfeuffer can you generate a secret for us that has write access over there?

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Add code to update website on release.

poshul committed:
update GHA action versions for Node 20

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
comment out tag creation of repo's we dont have access to right now

jpfeuffer commented:
I think this won't work, since the OpenMS GitHub application will not be able to tag across organizations I think. Maybe there is a way to allow the OpenMS application in GKN but I think we should maybe move the JS viewers to the OpenMS repo (same as the node wrapper and helper de.openms plugin).
GKN itself if I re-think now with more knowledege, does not need to be tagged. Compatible versions will be encoded in the de.openms plugin.

poshul commented:
Closing for now so that I can actually run the PR creation part.

poshul committed:
actually define outputs

poshul committed:
remove upload-url stuff

poshul committed:
Merge remote-tracking branch 'upstream/develop' into release/2.9.1-action

poshul committed:
tidy

poshul committed:
fix incorrectly escaped naming for upload artifact

poshul committed:
added creating a src archive and uploading it as part of the release

poshul committed:
fixed (hopefully)

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

actually upload the artifact

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

If there was any documentation as to what directory exclusions were relative to it would make this a hell of a lot easier.

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Update openms_ci_matrix_full.yml

poshul committed:
Merge remote-tracking branch 'upstream/develop' into release/2.9.1-action

poshul committed:
Merge branch 'release/2.9.1-action' of ssh://github.com/OpenMS/OpenMS into release/2.9.1-action

poshul committed:
add user.email and user.name

poshul committed:
moved the website update to another job

poshul committed:
create token for website

jpfeuffer committed:
Correct order to use token at checkout

poshul committed:
base64 encode relesae announcement so it doesn't get mangled

poshul committed:
fix base64 wrapping

poshul committed:
we used env.run_name in a lot of places where we should have used github.ref_name, they are the same when run automatically, but not when run manually

poshul committed:
fixed evaluation of nightly status for where to upload

poshul committed:
fix paths and variables for uploading.

poshul committed:
fixed links for latest

poshul committed:
redirect the upload paths to where they should live post-testing

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
Merge branch 'develop' into release/2.9.1-action

poshul committed:
update knime version to 5.2

poshul committed:
allow trailing text in release versions

poshul commented:
FYI openms-ci-full / build-and-test (ubuntu-22.04, g++, 11) is going to fail because I removed the version number override, and 2.9.1 doesn't actually have a changelog

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -674,3 +882,156 @@ jobs:
+              head: '${{ github.ref_name }}',
+              base: 'develop',
+              body: [
+                'DO NOT MERGE ME!!!!! This PR is auto-generated by',
```
I think this can have a different description now, right? The goal is to merge at some point right?

poshul committed:
remove DO NOT MERGE from release PR text.

jpfeuffer commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -436,21 +545,99 @@ jobs:
+            cat releaseTextHeader.txt $GITHUB_WORKSPACE/changelog.txt releaseTextFooter.txt > RELEASE_TEXT_GH.md
+            echo release_text_contents=`base64 -w0 RELEASE_TEXT_GH.md` >> $GITHUB_OUTPUT
+      
+      - name: run GHA release action.
```
Could you test this step? I can't find a release for 2.9.1.
Or do you delete them quickly always?

poshul replied:
Yeah, I've been deleting them.

jpfeuffer reviewed (approved):
Awesome thanks
